Add LLVM optimization and code generation backend

aengelke opened the pull request:
Implementation of an X86-64 assembly frontend for LLVM-IR and a corresponding interface for DBrew.

---

The build failures in Travis are currently caused by external dependencies exclusively:
- LLVM: GCC >= 5 - pedantic about enum value `1 << 31`
- LLVM: Clang 3.9 - does not like the LLVM 3.8 C++ headers...
- GLib: GCC >= 6 - pedantic about `__PRETTY_FUNCTION__`, old (5 years!) library, new compiler

aengelke committed:
Add LLVM optimization and code generation backend

aengelke committed:
Add .gitignore

weidendo commented:
Can you split that into multiple parts/commits/PRs, e.g. separate (1) llvm support code (2) interfacing with DBrew (3) adding tests for the interfacing (4) add the benchmark with tests ?
The points is that different parts may need different iterations to make it mergable.

aengelke committed:
ll: Fix invalid memory accesses

aengelke committed:
Reverse order of fall through and branch

This gives the precedence to the fall through branch.

aengelke committed:
Support multiple DBrew CBBs

To make life easier, the CBB got a new field generatorData, where the
generator (in this case the LL back-end) can store data temporarily.

aengelke committed:
Small refactoring of benchmark program

aengelke committed:
Integrate LL backend into build system

aengelke committed:
Improve Travis builds

* The Travis Docker systems are used, which means faster boot times
* LLVM 3.7 (which is required by the back-end) is now installed
* C++ is supported in the Makefile for the back-end

aengelke committed:
ll: Fix a few warnings

aengelke committed:
ll: Make target triple constant

aengelke committed:
Fix JS decoding test case

How did this even compile?

aengelke committed:
ll: Implement dispose methods

aengelke committed:
ll: Refactor stencil benchmark program

aengelke committed:
ll: Replace constant address with variables

There is a single global variable which is used as base address for all
accesses to absolute memory locations. It is initialized with the first
constant memory address found. In the JIT (ExecutionEngine) it is again
bound to the original location, so that the functionality is identical
but the LLVM IR does not contain addresses any more.

aengelke committed:
Allow test script to use LLVM without make

I'm not sure whether this is the best and simplest way to achieve this
without duplicating the test script.

aengelke committed:
Update tests to not include absolute addresses

The mov-scaling tests is now RIP-relative. For the constant-load test
only the expect file is removed. This means that the test case has to
run (i.e. not crash, exit code 0), but the output is not verified.

aengelke committed:
Delete test cases which depend on the new decoder

aengelke committed:
ll: Use vector instructions for fp values

On x86-64, floating-point values we use are always in vector registers.

aengelke committed:
ll: Implement heuristic to guess value type

When doing an arithmetic operation, we can improve the code by using
pointer arithmetics. However, if the value type is actually an integer,
the result will be worse. So, we need to guess whether a value is a
pointer or not just by looking at the value in the current basic block.

aengelke committed:
ll: Allow fixing of memory region

aengelke committed:
ll: Strip value names for Travis

This is the same version of LLVM but different names???

aengelke committed:
ll: Implement binding for LLVM intrinsics

Defining the intrinsics by hand is not really necessary as the C++ API
offers methods to access the directly.

aengelke committed:
ll: Use comparison for sign flag

Usually, LLVM can handle comparisons much better than bitwise
operations. The sign flag can also be checked by doing a comparison.

aengelke committed:
ll: Attempt to fix Travis build

aengelke committed:
ll: Disable unsafe pointer optimization by default

As these optimizations are not identical to the integer arithmetic, they
must be enabled separately. For example, the pointer arithmetic does not
support overflows but the result is undefined.

aengelke committed:
ll: Refine heuristic for values being a pointer

aengelke committed:
ll: Significantly improve stack handling

Previously, the stack offset was computed by subtracting the stack base
from the current stack pointer. The new stack pointer was computed via
add and sub instructions. This was good and much better than a naïve
implementation.However, for two pushes the optimizer got confused rather
quickly by too many inttoptr/sub/ptrtoint instructions.

Now, the stack pointer is adjusted via GEP instructions. The LLVM
optimizer is tuned for alias analysis on GEPs. As a consequence, now the
sequence ptrtoint/inttoptr will be emitted, which will be handled by the
optimizer. For testing purposes, a method which works using stack
offsets is still implemented, the results turned out to be equivalent.

Combined with unsafe pointer optimizations, also direct modifications of
the stack pointer (e.g. sub rsp, 0x20) will be handled well and
transformed into SSA registers.

aengelke committed:
ll: Handle global NULL pointer explicitly

aengelke committed:
ll: Handle IMUL with three operands

aengelke committed:
ll: Reduce vector register size to 128 bit

LLVM is apparently confused by an unused upper-half. We need to figure
out a different way to allow AVX extensions later.

aengelke committed:
ll: Pass alignment information to LLVM

This allows an instruction to specify the maximum alignment of memory
operands based on the operand size or a specific lower alignment.

However, x86 does only enforce alignment for SSE instructions but not
for general purpose instructions (like MOV or ADD). For these, the
alignment will be still set to maximum as we can expect that LLVM emits
general purpose instructions again for these memory operands.

aengelke committed:
ll: Fix benchmark program for LLVM cases

When the stencil pointer is null (which is the case for the native
implementation), the pointer must not be fixed.

aengelke committed:
ll: Remove unused field for stack

aengelke committed:
ll: Do not run vectorizer in first pass

The vectorizer is too aggressive and prohibits the elimination of
redundant instructions.

aengelke committed:
ll: Add option to use LLVM's overflow intrinsics

aengelke committed:
ll: Rename flag for unsafe pointer opts

aengelke committed:
ll: Strip dead prototypes

aengelke committed:
ll: Use pass manager builder

The optimization now relies on the LLVM pass manager builder, which is
tuned for C/C++ like languages. This has the effect that we only need to
enable vectorization (via C++ API) and only need two runs of the pass
manger now.

aengelke committed:
ll: Declare function from previous commit

aengelke committed:
ll: Benchmark LL fixed variant separately

aengelke committed:
ll: Minor cleanup of instruction file

aengelke committed:
ll: Fix handling of movs[sd] for memory sources

The upper part up to 127:32/64 is zeroes when loading from memory.
See Intel Vol. 2A 3-590 and Intel Vol. 2A 3-596.

aengelke committed:
ll: Improve storing of a full-sized vector

aengelke committed:
ll: Split operand and instruction handling

aengelke committed:
ll: Split registers and flags

Registers and flags are now separated and should be accessed via the
appropriate get/set functions.

aengelke committed:
ll: Access flag cache via a function

aengelke committed:
ll: Move jump handling to basic block

aengelke committed:
ll: Don't access basic block internals directly

aengelke committed:
ll: Restructure function class

aengelke committed:
ll: Allow specialization on function level

aengelke committed:
ll: Remove hacked variant of parameter fixing

aengelke committed:
ll: Use function type of base function

aengelke committed:
ll: Clone attributes from base function

aengelke committed:
ll: Move function typedef to correct header

aengelke committed:
ll: Register functions in the global context

aengelke committed:
ll: Implement rudimentary support for CALLs

aengelke committed:
ll: Implement MOVZX instruction

aengelke committed:
ll: Allow function declarations

aengelke committed:
ll: Improve function declarations

Now, if no symbol for the name exists and the name is not an LLVM
intrinsic, the function will be bound to the supplied address.

aengelke committed:
ll: Fix off-by-one error in decoder

aengelke committed:
ll: Add a few other test cases

aengelke committed:
ll: Allocate less basic blocks at the beginning

aengelke committed:
ll: Remove dangling test case

aengelke committed:
ll: Add test cases for pointer adjustments

aengelke committed:
ll: Huge improvements to the documentation

aengelke committed:
ll: Add Doxyfile

aengelke committed:
ll: Uncomment a few instructions

With the latest decoder changes, a few more instructions are
implemented. However, MOVHP[SD] is implemented *wrong*, since not the
lowest 64-bit of the register are handled but the higher 64-bits.
The correct value type is VT_128 for MOVHP[SD].

aengelke committed:
ll: Add test for MOV[LHU]PS

aengelke committed:
Travis Tampering

aengelke committed:
ll: Remove remaining C test cases

These aren't tested anyway due to platform specific differences of
different C compilers.

aengelke committed:
ll: Only use a fraction of LLVM libraries

aengelke committed:
ll: :zap: Basic support for function signatures

aengelke committed:
ll: Improve support for calls

aengelke committed:
ll: Better implementation of vector PXOR/XORP[SD]

aengelke committed:
ll: Support floating-point returns

aengelke committed:
ll: Set signature for declarations

aengelke committed:
ll: Fix XORP[SD] handling

LLVM does not support logical/bitwise operations on floating-point
values.

aengelke committed:
ll: Support void return type

aengelke committed:
ll: Fix benchmark program

aengelke committed:
ll: Allow functions to have private linkage

Functions with private linkage allow more optimization possibilities for
the inter-procedure optimizer.

aengelke committed:
Fix Makefile for parallel builds

aengelke committed:
ll: Fix specialization of void function

aengelke committed:
ll-benchmark: Performance improvements

The variants where two elements are computed at once now use start at an
even offset. Also, the first element, which must be preserved is not
stored in a register rather than on the stack to reduce memory accesses.

This also includes some sort of bug fix: no values are computed twice
any more.

aengelke committed:
ll: Allow to be less verbose about decoding

If verbose decoding is wanted, it can be still enabled via the DBrew
configuration.

aengelke committed:
ll-benchmark: Make interlines variable

This still happens at compile time for better optimizations

aengelke committed:
ll-benchmark: Add benchmark script

aengelke committed:
ll: Minor clean up

aengelke committed:
ll: Implement naive decoding

For comparison, it is now possible to use a naive decoding schema which
does not de-duplicate instructions.

aengelke committed:
More Travis...

aengelke committed:
ll: Try to use Clang for tests, too

aengelke committed:
ll: Decode test only in debug mode

This allows us to use Clang for the compilation, too.

On the other hand, I'm not quite sure whether testing multiple compilers
makes sense at all, since things should "just work" via the C standard..
...oh, wait we catch some Clang assembler bugs this way.

aengelke committed:
ll: Honor DBrew verbosity configuration in back-end

aengelke committed:
ll: Support UD2 instruction

Sometimes, GCC emits the ud2 instruction to mark branches as
unreachable. We can use this information here, too: we can mark the
branch as unreachable in LLVM.

aengelke committed:
ll: Fix for multiple DBrew breakages

It is not only DBrew which is breaking many times, now also the test
system itself starts breaking. I guess this is not the last time.

aengelke committed:
ll-benchmark: Set parameter count

aengelke committed:
ll: Remove -Waggregate-return warning

With the latest DBrew changes in the register structure design, this
will throw hundreds of warnings which are now unavoidable.

aengelke committed:
ll: Adjustments to DBrew's register type design

The improved design enforces multiple register files for different
register types (e.g. general purpose registers and XMM registers).

It also makes it necessary to handle the length of a floating-point
register manually, since XMM registers "always have a width of 128 bit".
Note that the general purpose registers, which in fact always have a
length of 64 bit, *do* have different widths, therefore this explicit
handling not necessary for those.

aengelke committed:
Update travis compiler versions

aengelke committed:
ll: Bump LLVM version to 3.8

aengelke committed:
Fix compiler warnings about enum comparisons

aengelke committed:
ll: Set flags for INC and DEC instructions

In some cases, GCC uses these instructions before a conditional branch.
Note that Intel discourages the use of INC and DEC instructions to
prevent unnecessary dependencies on the Carry Flag, which is preserved
for these instructions.

aengelke committed:
ll-benchmark: Use llvm-config in Makefile

aengelke committed:
ll: Support MOVHPD instruction

aengelke committed:
ll-benchmark: Add benchmark over lines

aengelke committed:
ll: Support 8-bit and 16-bit registers

This also includes the legacy register AH, CH, DH, BH.

aengelke committed:
ll: Move register write-back to separate function

aengelke committed:
ll: Simplify loading from integer registers

aengelke committed:
ll: Make function to build IR private

aengelke committed:
ll: Make instruction deduplication a runtime option

aengelke committed:
ll: Support fast math optimizations

Similar to -ffast-math, unsafe floating-point optimizations can be
enabled now.

aengelke committed:
ll: Fix handling of vector registers with MOVD/Q

aengelke committed:
ll: Fix handling of MOVHP[SD]

These instructions behave differently depending on the write direction.

aengelke committed:
ll: Restructure test directory

aengelke committed:
ll-benchmark: Refactor and merge kernels

There is now a single benchmark program which is capable of benchmarking
the element kernel and the line kernel as different configurations.

aengelke committed:
ll: Add header for public API

aengelke committed:
ll: Use shufflevector when zeroing upper part of register

aengelke committed:
ll: Fix operand storing in pop instruction

aengelke committed:
ll: Refactor loading of vector registers

aengelke committed:
ll: Improve documentation

aengelke committed:
ll: Simplify handling of stack instructions

aengelke committed:
ll: Unify handling of PUSH, POP and LEAVE

aengelke committed:
ll: Move pointer type to global offset function

aengelke committed:
ll: Minor refactoring of vector element loading

aengelke committed:
ll: Avoid GEP from null pointer

Instead of a using NULL as base pointer, we use inttoptr instead.

aengelke committed:
ll: Fix invalid operand errors from attaching metadata

When setting metadata for constant expressions, the operator changes to
<invalid operand>. Circumvent this by setting metadata only for real
instructions and not for constants.

aengelke committed:
ll: Running the pass pipeline once suffices

aengelke committed:
ll: Test LEAVE instruction

aengelke committed:
ll: Allow to force loop unrolling

aengelke committed:
ll: Adjust fixation buffer size

aengelke committed:
ll: Fix warning about aggregate return

aengelke committed:
ll: Add Sparse-Matrix-Vector example

aengelke committed:
ll: Fix warnings about instruction type enum

aengelke committed:
Travis Experiment

aengelke committed:
ll: Introduce explicit facets

:tada: - part 1

aengelke committed:
ll: Use facets for loading registers

:tada: - part 2

It is noticable that the IR quality of the stencil/double test case
improves singificantly.

aengelke committed:
ll: Use facets for memory operands

:tada: - part 3

This allows us to remove the pointer detection heuristics.

aengelke committed:
ll: Provide an integer facet to the LEA instruction

:tada: - part 4

LEA is heavily used for integer operations as well. Additionally, on we
now rename the register if the full register is moved keeping all facets
in place and allowing further optimizations.

Additionally, a (currently unused) function to zero all facets is added.

aengelke commented:
:tada: :tada: :tada: Now, register facets are implemented explicitly and remove a **LOT** of overhead in more  complex code (e.g. the benchmark). The result of the LLVM identity transformation is now as fast (and sometimes even faster) as the original code.

aengelke committed:
ll: Add basic support for indirect calls

If the value of a register is a known constant, the call may succeed.

aengelke committed:
ll: Add some more SSE instructions

aengelke committed:
ll: Split instruction handling over multiple files

:zap: This introduces descriptors to additionally remove redundant code.

aengelke committed:
ll: Implement MOVDQA/MOVDQU

aengelke committed:
ll: Initial support for AVX 256-bit registers

:tada: - part 5

Now, it is possible to enable support for 256-bit registers without
large overhead. Still, the instruction semantics are not adjusted yet.

aengelke committed:
ll: Implement correct partial register handling for AVX

:tada: - part 6

aengelke committed:
ll: Do not hard-code DBrew decoder

This allows the integration of other decoders with a small wrapper
function to convert to the used instruction description.

aengelke committed:
ll: Make SSE instruction handling more robust

aengelke committed:
ll: Use small code model for more efficient code

aengelke committed:
ll-benchmark: Enable floating-point optimizations

aengelke committed:
ll-examples: Fix for recent changes

aengelke committed:
ll: Minor refactoring of register stores

aengelke committed:
ll: Attach register information to PHI nodes

aengelke committed:
ll: Split out registers in separate file

aengelke committed:
ll: Run some light-weight passes earlier

This can significantly reduce the compile time.

aengelke committed:
ll: Compile with -pedantic

aengelke committed:
ll: Support MUL and single-operand IMUL

aengelke committed:
ll: Add function to dump IR of a single function

aengelke committed:
ll: Support more operand sizes for LEA

aengelke committed:
ll: Handle pointer returns correctly

aengelke committed:
ll: Expose computation of individual flags

aengelke committed:
ll: Compute flags for NEG

aengelke committed:
ll: Set flags for MUL and IMUL

aengelke committed:
ll: Support PUSHFQ for status flags

All other flags are set to zero.

aengelke committed:
ll: Minor updates to comments

aengelke committed:
ll: Fix tests

aengelke committed:
ll: Major restructuring of tests

Correctness tests are now done via GLib + GTester in one binary. This
gives much better test times (~1s) and also increases maintainability,
as changes to the IR have significantly less impact on the tests.

There are now two types of tests:
- Correctness Tests: A function is transformed to LLVM IR and executed
  with 100 random inputs, the result is then compared with the original
  function. This also works for status flags and replaces most test
  cases. Note that this only ensures the correctness of the
  transformation but not the optimality. For this, we (still) have...
- Optimization Tests: This is basically the old testing strategy which
  compares the output of the optimized LLVM-IR with an expectation.

aengelke committed:
ll: Remove comments from Doxyfile

aengelke committed:
ll: Also compile C++ with -O2

aengelke committed:
ll: Set some flags for shift instructions

This is still incomplete, but should be enough to avoid bugs caused by
GCC using the result of the zero flag.

aengelke committed:
ll: Allow to pass options to LLVM

aengelke committed:
ll: Fix bug with long constant addresses

Previously, constant offsets were cut after 32-bits, which leads to
incorrect memory addresses if the distance to the base pointer is >4GiB.

This also extends the tests by adding an "MMAP" flag, which causes
read-only memory to be mapped at known addresses.

aengelke committed:
ll: Refactor constant pointer construction

aengelke committed:
ll: Another commit, because we love Clang
